File: apps/php/design-pattern/src/TemplateMethod/HtmlReport.php
<?php

declare(strict_types=1);

namespace DesignPattern\TemplateMethod;

class HtmlReport extends Report
{
    protected function outputStart(): array
    {
        return ['<html>'];
    }

    protected function outputHead(): array
    {
        return [' <head>', " <title>{$this->title}</title>", ' </head>'];
    }

    protected function outputBodyStart(): array
    {
        return ['<body>'];
    }

    protected function outputLine(string $line): array
    {
        return [" <p>{$line}</p>"];
    }

    protected function outputBodyEnd(): array
    {
        return ['</body>'];
    }

    protected function outputEnd(): array
    {
        return ['</html>'];
    }
}

File: apps/java/design-pattern/src/main/java/pattern/templatemethod/HtmlReport.java
package pattern.templatemethod;

/**
 * HTML 形式のレポート出力
 */
public class HtmlReport extends Report {

    @Override
    protected void outputStart(StringBuilder sb) {
        sb.append("<html>\n");
    }

    @Override
    protected void outputHead(StringBuilder sb) {
        sb.append(" <head>\n");
        sb.append(" <title>").append(title).append("</title>\n");
        sb.append(" </head>\n");
    }

    @Override
    protected void outputBodyStart(StringBuilder sb) {
        sb.append("<body>\n");
    }

    @Override
    protected void outputLine(StringBuilder sb, String line) {
        sb.append(" <p>").append(line).append("</p>\n");
    }

    @Override
    protected void outputBodyEnd(StringBuilder sb) {
        sb.append("</body>\n");
    }

    @Override
    protected void outputEnd(StringBuilder sb) {
        sb.append("</html>\n");
    }
}
